File: 백준/Gold/12919. A와 B 2/A와 B 2.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
#문자열 뒤에 A
#B추가 후 뒤집기

def dfs(target):
    global flag
    if len(S)==len(target):
        if S == target:
            flag=1
        return
    if target[-1]=='A':
        target.pop()
        dfs(target)
        target.append('A')

    if target[0]=='B':
        target.reverse()
        target.pop()
        dfs(target)
        target.append('B')
        target.reverse()
S = list(map(str, input().strip('\n')))
T = list(map(str, input().strip('\n')))
flag = 0
dfs(T)

print(flag)

File: 백준/Gold/12919. A와 B 2/A와 B 2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
	flag   int
	start  []string
)

func reverse(s []string) []string {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
	return s
}
func dfs(target []string) {
	if len(start) == len(target) {
		if strings.Join(start, "") == strings.Join(target, "") {
			flag = 1
		}
		return
	}
	if target[len(target)-1] == "A" {
		target = target[:len(target)-1]
		dfs(target)
		target = append(target, "A")
	}
	if target[0] == "B" {
		target = reverse(target)
		target = target[:len(target)-1]
		dfs(target)
		target = append(target, "B")
		target = reverse(target)
	}

}

func main() {
	defer writer.Flush()
	var s, t string
	fmt.Fscanf(reader, "%s\n%s\n", &s, &t)
	target := strings.Split(t, "")
	start = strings.Split(s, "")
	dfs(target)
	fmt.Fprintf(writer, "%d\n", flag)
}
